src/i3_json_parser/i3_json_parser-utility.cpp

```diff
@@ -271,5 +271,18 @@ i3_containers::node i3_json_parser::extract_tree(const rapidjson::Value& a_json_
         node.floating_nodes.push_back(extract_tree(node_object));
     }
 
+    if (a_json_object.HasMember("marks") && !a_json_object["marks"].IsNull())
+    {
+        auto& json_marks = a_json_object["marks"];
+        assert(json_marks.IsArray());
+        auto & marks = node.marks;
+        marks.reserve(json_marks.GetArray().Size());
+        for (const auto& mark : json_marks.GetArray())
+        {
+            assert(mark.IsString());
+            marks.push_back(mark.GetString());
+        }
+    }
+
```

Comment:
Can you please move this code to `extract_marks` function (in anonymous namespace, like the rest), [as suggested](https://github.com/Iskustvo/i3-ipcpp/pull/4#discussion_r475138686) in the original PR?
Then we can bundle that function call together with other extractions on top of this function, above the recursive sub-tree parsing.

Btw, judging from my very limited testing and this [upstream PR](https://github.com/i3/i3/pull/3970), `marks` are now part of every node/container response, so we should be fine just asserting that we have (possibly empty) array value.